fix(typeEvlauator): handle rest on object splat operations

sgulseth opened the pull request:
When splatting an object we didn't map over the rest property, if it's an object we should append the attributes. If `rest` is unknown we should treat the entire object as unknown.

Fixes https://github.com/sanity-io/sanity/issues/6555

judofyr commented:
Something looks a bit sketchy.

Here we invoke the mapper multiple times: https://github.com/sanity-io/groq-js/blob/15ec7ac20b00bc3f70ed72c56f8ddcae561930b0/src/typeEvaluator/typeEvaluate.ts#L124-L128

But here we assume that it's only being called once per `name`: https://github.com/sanity-io/groq-js/blob/15ec7ac20b00bc3f70ed72c56f8ddcae561930b0/src/typeEvaluator/typeEvaluate.ts#L165-L167

`mapObjectSplat` also only handles `union` and `object` and ignores `unknown`. It seems that if you now splat something which we can't determine it ends up being ignored? 🤔 

I'm not 100% sure what the right approach here is though. Maybe more usage of `mapConcrete`?

sgulseth commented:
yess, this should probably be a concrete type with mapConcrete, will fix. Didn't think much about fixing other bugs than the one referenced 😅

sgulseth commented:
@judofyr did another pass on comparing with mapped concrete type node

judofyr commented on src/typeEvaluator/typeEvaluate.ts:
```diff
@@ -119,18 +119,40 @@ function handleDerefNode(node: DerefNode, scope: Scope): TypeNode {
+  const attributes: Record<string, ObjectAttribute> = {}
+
+  const concreteNode = mapConcrete(node, scope, (node) => node)
+  if (concreteNode.type === 'object') {
```
Hmm. Doesn't this mean that if this returns `false` we end up returning `{type: 'object', attributes: {}}` below? What about we instead wrap all of this in `mapConcrete`?

```ts
return mapConcrete(node, scope, (node) => {
  if (node.type !== 'object') return {type: 'unknown'}
  const attributes: Record<string, ObjectAttribute> = {}
  // same same…
})
```

sgulseth replied:
I did another pass, moved the handling of the object splats out in its own methods

judofyr commented on src/typeEvaluator/typeEvaluate.ts:
```diff
@@ -145,44 +230,29 @@ function handleObjectNode(node: ObjectNode, scope: Scope) {
-        })
+      const result = handleObjectConditionalSplatNode(attr, scope)
+      $trace('object.conditional.splat.result %O', result)
+      if (result.type !== 'object') {
+        return result
+      }
```
I don't think this makes sense: If the splatted node is _not_ an object then we return that as the result of the whole object?

In the following query:

```groq
*[_type == "author"][0] {
    _type == "author" => select(age >= 18 => {lastname}, {firstname}),
    _id,
    _type
  }
```

Then we now end up generating:

```js
{
  "type": "union",
  "of": Array [
    Object {
      "type": "object",
      "attributes": Object {
        "firstname": Object {
          "type": "objectAttribute",
          "value": Object {
            "type": "string",
          },
        },
      },
    },
    Object {
      "type": "object",
      "attributes": Object {
        "lastname": Object {
          "type": "objectAttribute",
          "value": Object {
            "type": "string",
          },
        },
      },
    },
    Object {
      "type": "null",
    },
  ],
}
```

Because the "conditional splat" turns into a `union` and thus we hit this case.

judofyr commented on src/typeEvaluator/typeEvaluate.ts:
```diff
@@ -117,20 +119,103 @@ function handleDerefNode(node: DerefNode, scope: Scope): TypeNode {
+  return mapConcrete(value, scope, (node) => {
+    const attributes: Record<string, ObjectAttribute> = {}
+    const mapped = mapObjectSplat(node, scope, (attribute) => attribute)
+    if (mapped.type != 'object') {
+      return mapped
+    }
+    for (const name in mapped.attributes) {
+      if (!mapped.attributes.hasOwnProperty(name)) {
+        continue
+      }
+      attributes[name] = mapped.attributes[name]
+    }
+    return {type: 'object', attributes}
```
This code has just basically takes a `type: "object"` and then makes a copy of it without `rest` or `dereferencesTo`. That doesn't look right to me.

judofyr commented on src/typeEvaluator/typeEvaluate.ts:
```diff
@@ -117,20 +119,103 @@ function handleDerefNode(node: DerefNode, scope: Scope): TypeNode {
+    if (concreteRest.type === 'unknown') {
+      return {type: 'unknown'}
+    }
+    if (concreteRest.type !== 'object') {
+      return {type: 'null'}
```
This means that if `concreteRest.type` is "union" we end up returning null?

judofyr commented:
It's a bit unclear to me what the intended behavior of handling objects with unions and inlines are at the moment. If there's union in any of `ObjectSplat` there's multiple approaches:

- We can do the the "cross product": E.g. if there's 3 different object splats which all have 3 unions we end up with a top-level union of 9 different alternatives.
- We can choose that this is just always `{type: "unknown"}`.
- We can take the _last_ union of objects, change all of them to have `rest: unknown`. That way we get typings for the "last" set of attributes, but not the "earlier" ones.

Maybe we can add some more fine-grained tests using `overrideTypeForNode` which demonstrates how it will behave in various cases?

judofyr commented:
How about creating some tests like this (maybe turn it into a table-based one?):

```ts
const nodeWithType = (type: TypeNode) => {
  const expr: ExprNode = {type: 'Value', value: null}
  overrideTypeForNode(expr, type)
  return expr
}

t.test('objects', (t) => {
  // test 1
  typeEvaluate(
    {
      type: 'Object',
      attributes: [
        {type: 'ObjectAttributeValue', name: 'foo', value: nodeWithType({type: 'string'})},
      ],
    },
    [],
  )

  // test 2
  typeEvaluate(
    {
      type: 'Object',
      attributes: [
        {type: 'ObjectAttributeValue', name: 'foo', value: nodeWithType({type: 'string'})},
        {type: 'ObjectAttributeValue', name: 'foo', value: nodeWithType({type: 'number'})},
      ],
    },
    [],
  )

  // test 3
  typeEvaluate(
    {
      type: 'Object',
      attributes: [
        {type: 'ObjectAttributeValue', name: 'foo', value: nodeWithType({type: 'string'})},
        {type: 'ObjectSplat', value: nodeWithType({type: 'unknown'})},
      ],
    },
    [],
  )
})
```

sgulseth commented:
I've addressed your feedback(100% agree), we should now be handling unions by always using mapConcrete. I also moved resolveInline into a separate function so we can be explicit about resolving inline nodes.

sgulseth commented on test/typeEvaluate.test.ts:
```diff
@@ -2406,6 +2430,800 @@ t.test('function: math::*', (t) => {
+        }),
+      },
+    ],
+    expects: unionOf(
+      {
+        type: 'object',
+        attributes: {
+          foo: {type: 'objectAttribute', value: {type: 'boolean'}},
+        },
+      },
+      {
+        type: 'unknown',
+      },
+    ),
```
We could short circuit so that if any of the returned elements are unknown we return unknown. Guess this could happen if you have a schema where a field is either an object or and array of something, not sure that's very realistic, though it could be during a migration for example. 

If you have multiple attributes in the object, you could simplify the type check by just asserting one value and typescript would assume it's not the unknown object 🤔 

Personally no strong opinions

sgulseth commented:
Pushed my latest changes, there's optimizations to be done in the object handler itself(probably way too many loops 😅), but if it's correct I can give optimizations another passs

judofyr reviewed (approved):
I've mainly focused on the tests and they all look sensible to me.

(I haven't looked much at the implementation. Let me know if there's anything particular there you'd like me to review. Otherwise I'm pretty happy doing a "test-first" approach for this code anyway.)
